### PR TITLE
Rework StreamBuilder to avoid unnecessary transcoding

### LINKED ISSUE #6149: [bug] Unnecessarily transcoding vp9 in Firefox
**Describe the bug**
Jellyfin on web client is transcoding during playback of vp9 video when vp9 is supported by Firefox and my computer. 

**System (please complete the following information):**
 - OS: Ubuntu Server 20.04.2
 - Virtualization: Docker
 - Clients: Browser
 - Browser: Firefox 88.0.1 (on Ubuntu 21.04)
 - Jellyfin Version: 10.7.5
 - Playback: Transcode
 - Installed Plugins: none? manually installed none, default plugins (AudioDB, MusicBrainz, OMDb)
 - Reverse Proxy: caddy
 - Base URL: none
 - Networking: Host
 - Storage: local

**To Reproduce**

1. Play vp9 video on firefox on ubuntu. Open playback information.

**Expected behavior**
Transcoding only audio, or nothing (idk if aac is supported on firefox)

**Server Logs**

**FFmpeg Logs**

**Browser Console Logs**

**Screenshots**

Screenshot from 2021-06-04 03-52-16 censored

**Additional context**
Video was downloaded with youtube-dl, not sure how it handles metadata if that is relevant? Playback was also occasionally dropping frames, I assume transcoding may have something to do with it.

If you need me to add server logs, ffmpeg logs, or browser console logs, I am willing, but do not know how.

### LINKED ISSUE #6450: Avoid unnecessary transcoding by only transcoding incompatible streams
**Describe the bug**
This is a meta-bug that causes #6149, #6449 (TODO: look for other bugs relating to this).

If only some of the streams (or the container itself) is unsupported, only transcode the unsupported streams and just copy/mux the other streams.

**System (please complete the following information):**
 - OS: All
 - Virtualization: Any
 - Clients: All
 - Browser: All
 - Jellyfin Version: 10.7.6
 - Playback: Transcode
 - Hardware Acceleration: All
 - Installed Plugins: All
 - Reverse Proxy: All
 - Base URL: All
 - Networking: All
 - Storage: All

**To Reproduce**

1. Attempt to play a video to a client that supports the video codec but not the audio or container. For example, play a (MKV, VP9, AAC) video on Firefox.
2. Notice that transcoding occurs unnecessarily for the video stream as well as the audio stream.

**Expected behavior**

Only the unsupported the streams should be transcoded and supported streams should be passed through.

**Additional context**

This exposes a deeper issue with some of the transcoding logic that needs to be fixed.

Refer to `MediaBrowser.Model.Dlna.Streambuilder.BuildVideoItem()` for a starting reference point for this bug.

For example, something similar to this logic should be used to pick a target transcode output format:

```csharp
// Can't direct play, find the transcoding profile
TranscodingProfile transcodingProfile = null;
int bestMaxAudioChannels = 0;
foreach (var i in options.Profile.TranscodingProfiles)
{
    if (i.Type == playlistItem.MediaType && i.Context == options.Context)
    {
        int curMaxAudioChannels = 0;
        int.TryParse(i.MaxAudioChannels, out curMaxAudioChannels);

        if (transcodingProfile == null)
        {
            transcodingProfile = i;
        }
        else if (i.VideoCodec == NormalizeSourceCodecToTranscodeCodec(item.VideoStream.Codec) && transcodingProfile.VideoCodec != NormalizeSourceCodecToTranscodeCodec(item.VideoStream.Codec))
        {
            // Container supports original codec, so pick this profile instead.
            transcodingProfile = i;
        }
        else if (curMaxAudioChannels > bestMaxAudioChannels)
        {
            // Container supports more audio channels output.
            transcodingProfile = i;
        }
        // Probably add logic about containers that support embedded subtitles.
    }
}
```

Then, there should be some indication, perhaps in `MediaBrowser.Model.Dlna.StreamInfo`, to indicate whether to pass through the audio or video streams to the output container, and the transcoding logic should selectively transcode.

### LINKED ISSUE #6540: Cannot access jellyfin outside localhost.
**Describe the bug**
I cannot access the jellyfin webUI or from a third party app or official iOS app locally, only the host can open the webUI.

**System (please complete the following information):**
 - OS: OpenSUSE
 - Virtualization: Docker
 - Clients: Firefox, Firefox on iOS, MrMC on AppleTV and official iOS app
 - Browser: Firefox 91, Firefox Daylight 36
 - Jellyfin Version: 10.7.7 
 - FFmpeg Version: excuse me?
 - Playback: irrelevant 
 - Hardware Acceleration: default
 - Installed Plugins: none
 - Reverse Proxy: none
 - Base URL: none
 - Networking: Host, apparently
 - Storage: local

**To Reproduce**

1. Go to [internal IP]:8096
2. Error: "Unable to connect Firefox can't stablish a connection to the server at [internal IP]:8096 

**Expected behavior**
Load the WebUI or connect to the server on the iOS app or the tvOS app.

**Server Logs**
https://gist.github.com/lavadrop/9fa445c0ed8d3bb99daf6aa5950594da

**Additional context**
Installed the jellyfin docker image per the main download page code at jellyfin.org (stable)

### PR DESCRIPTION
**Pull request by eyezak**
**Changes**
Adds preferred matching of video and audio streams to both DirectPlay and Transcode profiles. Multiple candidate profiles are considered after codec resolution, with preference to copy-codec. In the case of compatible video streams, remuxing is preferred to transcoding audio. In the cases of no compatible video streams,  multiple equivalent candidate profiles and !Allow*Copy - preference is given to the client-supplied profile order to support prefer fMP4/ts streams user-setting. In contrast to the approach of https://github.com/jellyfin/jellyfin/pull/6454, PlaybackInfo based apis now prefer transcoding to direct streaming to allow direct-stream equivalent transcoding to HLS (ts or fMP4).
MediaInfoHelper has been adapted to only call StreamBuilder once.

Converts TranscodeReason to Flags and serialize into strings in models for simplicity.
Adds a suite of tests for StreamBuilder based on real and imagined device profiles.

Currently testing on Firefox/Chrome/Safari with av1,divx,h264,hevc,msmp4,xvid x aac,ac3,eac3,mp3,flac,opus,vorbis x mp4,mkv,avi,webm encodes of the Sintel 2048 trailer. Verified HEVC10 and HEVC on safari with remux and transcode as direct-stream with web-player modifications (https://github.com/jellyfin/jellyfin-web/pull/3442) .

**Issues**
fixes #6450
fixes #6149
alternative approach to #6454 
Potentially others i have not validated yet

### CODE REVIEW

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 1]
```diff
@@ -15,6 +15,12 @@ namespace MediaBrowser.Model.Dlna
 {
     public class StreamBuilder
     {
+        // Aliases
+        public static readonly TranscodeReason ContainerReasons = TranscodeReason.ContainerNotSupported | TranscodeReason.ContainerBitrateExceedsLimit;
+        public static readonly TranscodeReason AudioReasons = TranscodeReason.AudioCodecNotSupported | TranscodeReason.AudioBitrateNotSupported | TranscodeReason.AudioChannelsNotSupported | TranscodeReason.AudioProfileNotSupported | TranscodeReason.AudioSampleRateNotSupported | TranscodeReason.SecondaryAudioNotSupported | TranscodeReason.AudioBitDepthNotSupported | TranscodeReason.AudioIsExternal;
+        public static readonly TranscodeReason VideoReasons = TranscodeReason.VideoCodecNotSupported | TranscodeReason.VideoResolutionNotSupported | TranscodeReason.AnamorphicVideoNotSupported | TranscodeReason.InterlacedVideoNotSupported | TranscodeReason.VideoBitDepthNotSupported | TranscodeReason.VideoBitrateNotSupported | TranscodeReason.VideoFramerateNotSupported | TranscodeReason.VideoLevelNotSupported | TranscodeReason.RefFramesNotSupported;
+        private static readonly TranscodeReason DirectStreamReasons = AudioReasons | TranscodeReason.ContainerNotSupported;
```
Comment (crobibero) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:22]:
```suggestion
        internal const TranscodeReason ContainerReasons = TranscodeReason.ContainerNotSupported | TranscodeReason.ContainerBitrateExceedsLimit;
        internal const TranscodeReason AudioReasons = TranscodeReason.AudioCodecNotSupported | TranscodeReason.AudioBitrateNotSupported | TranscodeReason.AudioChannelsNotSupported | TranscodeReason.AudioProfileNotSupported | TranscodeReason.AudioSampleRateNotSupported | TranscodeReason.SecondaryAudioNotSupported | TranscodeReason.AudioBitDepthNotSupported | TranscodeReason.AudioIsExternal;
        internal const TranscodeReason VideoReasons = TranscodeReason.VideoCodecNotSupported | TranscodeReason.VideoResolutionNotSupported | TranscodeReason.AnamorphicVideoNotSupported | TranscodeReason.InterlacedVideoNotSupported | TranscodeReason.VideoBitDepthNotSupported | TranscodeReason.VideoBitrateNotSupported | TranscodeReason.VideoFramerateNotSupported | TranscodeReason.VideoLevelNotSupported | TranscodeReason.RefFramesNotSupported;
        internal const TranscodeReason DirectStreamReasons = AudioReasons | TranscodeReason.ContainerNotSupported;
```

And add the following to `MediaBrowser.Model/Properties/AssemblyInfo.cs` (under `[assembly: InternalsVisibleTo("Jellyfin.Model.Tests")]`

```c#
[assembly: InternalsVisibleTo("Jellyfin.Dlna.Tests")]
```

File: MediaBrowser.Model/Session/TranscodeReason.cs [hunk 2]
```diff
@@ -1,32 +1,46 @@
 #pragma warning disable CS1591
 
+using System;
+
 namespace MediaBrowser.Model.Session
 {
+    [Flags]
```
Comment (crobibero) [inline MediaBrowser.Model/Session/TranscodeReason.cs:7]: Just as a note- changing this enum to `Flags` changed the integer representation of each value. This shouldn't matter as the API returns the string.

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.Model/Session/TranscodeReason.cs (hunk 2 above)
Comment (eyezak) [reply MediaBrowser.Model/Session/TranscodeReason.cs:7]: yeah, i saw that the api was string based and I chose to change the values and have the sort-order be meaningful as well

**Comment by eyezak** [comment]
I added another commit. I found during rebase I missed a transcode-matching case - added tests and updated SafariNext test device profile to match https://github.com/jellyfin/jellyfin-web/pull/3442

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 3]
```diff
@@ -689,189 +628,299 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
...
+                int? numVideoStreams = item.GetStreamCount(MediaStreamType.Video);
+
+                transcodingProfiles = transcodingProfiles.ToLookup(transcodingProfile =>
                 {
-                    transcodingProfile = i;
-                    break;
-                }
+                    var videoCodecs = ContainerProfile.SplitValue(transcodingProfile.VideoCodec);
+
+                    if (ContainerProfile.ContainsContainer(videoCodecs, item.VideoStream.Codec)) {
```
Comment (crobibero) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:753]:
```suggestion
                    if (ContainerProfile.ContainsContainer(videoCodecs, item.VideoStream.Codec))
                    {
```

File: MediaBrowser.Model/Session/TranscodeReasonExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        /// <summary>
+        /// Serializes a TranscodeReason into an array of individual TranscodeReason bits.
+        /// </summary>
+        /// <param name="reasons">The <see cref="TranscodeReason"/> enumeration.</param>
+        /// <returns>Array of <c>TranscodeReason</c>.</returns>
+        public static TranscodeReason[] ToArray(this TranscodeReason reasons)
+        {
+            return _values.Where(r => r != 0 && reasons.HasFlag(r)).ToArray();
+        }
+    }
```
Comment (Bond-009) [inline MediaBrowser.Model/Session/TranscodeReasonExtensions.cs:33]: `TranscodeReason.ToString()` should do exactly this afaik?

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 5]
```diff
@@ -689,189 +628,300 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
...
+                if (playlistItem.AudioBitrate.HasValue)
+                {
+                    availableBitrateForVideo -= playlistItem.AudioBitrate.Value;
                 }
-            }
 
-            playlistItem.TranscodeReasons = transcodeReasons.ToArray();
+                // Make sure the video bitrate is lower than bitrate settings but at least 64k
+                long currentValue = playlistItem.VideoBitrate ?? availableBitrateForVideo;
+                var longBitrate = Math.Max(Math.Min(availableBitrateForVideo, currentValue), 64000);
```
Comment (Bond-009) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:913]: `Math.Clamp` exists

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 6]
```diff
@@ -689,189 +628,300 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
...
+                playlistItem.VideoBitrate = longBitrate >= int.MaxValue ? int.MaxValue : Convert.ToInt32(longBitrate);
```
Comment (Bond-009) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:914]: Not sure about the `Convert.ToInt32`, only thing it does it trow if `longBitrate < int.MinValue`

File: src/Jellyfin.Extensions/Json/JsonDefaults.cs [hunk 7]
```diff
@@ -36,6 +36,7 @@ public static class JsonDefaults
                 new JsonGuidConverter(),
                 new JsonNullableGuidConverter(),
                 new JsonVersionConverter(),
+                new JsonFlagEnumConverterFactory(),
```
Comment (Bond-009) [inline src/Jellyfin.Extensions/Json/JsonDefaults.cs:39]: Does this work as expected? When is `JsonStringEnumConverter` used now?

**Review by Bond-009 — DISMISSED** [review]

File: Jellyfin.Api/Controllers/UniversalAudioController.cs [hunk 8]
```diff
@@ -223,7 +224,7 @@ public class UniversalAudioController : BaseJellyfinApiController
                     DeInterlace = false,
                     RequireNonAnamorphic = false,
                     EnableMpegtsM2TsMode = false,
-                    TranscodeReasons = mediaSource.TranscodeReasons == null ? null : string.Join(',', mediaSource.TranscodeReasons.Select(i => i.ToString()).ToArray()),
+                    TranscodeReasons = mediaSource.TranscodeReasons == 0 ? null : mediaSource.TranscodeReasons.ToString(),
```
[the remark this answers is not in the document]
Comment (crobibero) [reply Jellyfin.Api/Controllers/UniversalAudioController.cs:227]: Originally did have `None`, but the JsonConverter and ToString would always list None as a value....

File: src/Jellyfin.Extensions/Json/JsonDefaults.cs (hunk 7 above)
Comment (crobibero) [reply src/Jellyfin.Extensions/Json/JsonDefaults.cs:39]: `JsonStringEnumConverter` is used for any enum that is not `[Flags]`

File: Jellyfin.Api/Controllers/UniversalAudioController.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Bond-009) [reply Jellyfin.Api/Controllers/UniversalAudioController.cs:227]: That's stupid :/

File: MediaBrowser.Controller/MediaEncoding/EncodingJobInfo.cs [hunk 9]
```diff
@@ -34,25 +35,23 @@ public EncodingJobInfo(TranscodingJobType jobType)
             SupportedSubtitleCodecs = Array.Empty<string>();
         }
 
-        public TranscodeReason[] TranscodeReasons
+        public TranscodeReason TranscodeReason
```
[the remark this answers is not in the document]
Comment (crobibero) [reply MediaBrowser.Controller/MediaEncoding/EncodingJobInfo.cs:38]: That's an API breaking change

**Review by joshuaboniface — APPROVED** [review]

**Comment by nyanmisaka** [comment]
```
[16:41:17] [ERR] [41] Jellyfin.Server.Middleware.ExceptionMiddleware: Error processing request. URL POST /Items/f904cec75fb51710ef3db98195be9787/PlaybackInfo.
System.ArgumentException: '14616000' cannot be greater than 64000.
   at System.Math.ThrowMinMaxException[T](T min, T max)
   at MediaBrowser.Model.Dlna.StreamBuilder.BuildStreamVideoItem(StreamInfo playlistItem, VideoOptions options, MediaSourceInfo item, MediaStream videoStream, MediaStream audioStream, IEnumerable`1 candidateAudioStreams, String container, String videoCodec, String audioCodec)
   at MediaBrowser.Model.Dlna.StreamBuilder.BuildVideoItem(MediaSourceInfo item, VideoOptions options)
   at MediaBrowser.Model.Dlna.StreamBuilder.BuildVideoItem(VideoOptions options)
   at Jellyfin.Api.Helpers.MediaInfoHelper.SetDeviceSpecificData(BaseItem item, MediaSourceInfo mediaSource, DeviceProfile profile, AuthorizationInfo auth, Nullable`1 maxBitrate, Int64 startTimeTicks, String mediaSourceId, Nullable`1 audioStreamIndex, Nullable`1 subtitleStreamIndex, Nullable`1 maxAudioChannels, String playSessionId, Guid userId, Boolean enableDirectPlay, Boolean enableDirectStream, Boolean enableTranscoding, Boolean allowVideoStreamCopy, Boolean allowAudioStreamCopy, IPAddress ipAddress)
   at Jellyfin.Api.Controllers.MediaInfoController.GetPostedPlaybackInfo(Guid itemId, Nullable`1 userId, Nullable`1 maxStreamingBitrate, Nullable`1 startTimeTicks, Nullable`1 audioStreamIndex, Nullable`1 subtitleStreamIndex, Nullable`1 maxAudioChannels, String mediaSourceId, String liveStreamId, Nullable`1 autoOpenLiveStream, Nullable`1 enableDirectPlay, Nullable`1 enableDirectStream, Nullable`1 enableTranscoding, Nullable`1 allowVideoStreamCopy, Nullable`1 allowAudioStreamCopy, PlaybackInfoDto playbackInfoDto)
   at lambda_method1192(Closure , Object )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
```

**Review by nyanmisaka — APPROVED** [review]

**Comment by nyanmisaka** [comment]
@eyezak Hi! I found that after this PR, media can be played directly on Chrome regardless of the resolution and bitrate limitations set by jellyfin-web. Any ideas on this?

**Comment by dmitrylyzo** [comment]
https://github.com/jellyfin/jellyfin/blob/e1527857e931527f68b9705e248bd7bc83c62940/MediaBrowser.Model/Dlna/StreamBuilder.cs#L617-L618
`GetBitrateForDirectPlayCheck` doesn't use the user-defined bitrate limit.
https://github.com/jellyfin/jellyfin/blob/e1527857e931527f68b9705e248bd7bc83c62940/MediaBrowser.Model/Dlna/StreamBuilder.cs#L367-L370
That's why direct-playable video is not transcoded according to the user-defined bitrate limit.

**Pull request merged**